usr/man/man/usr/share/man/man1/time.1
'\" te
.\"  Copyright (c) 1992, X/Open Company Limited  All Rights Reserved  Portions Copyright (c) 1995, Sun Microsystems, Inc.  All Rights Reserved
.\"  Sun Microsystems, Inc. gratefully acknowledges The Open Group for permission to reproduce portions of its copyrighted documentation. Original documentation from The Open Group can be obtained online at  http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group, have given us permission to reprint portions of their documentation. In the following statement, the phrase "this text" refers to portions of the system documentation. Portions of this text are reprinted and reproduced in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1, 2004 Edition, Standard for Information Technology -- Portable Operating System Interface (POSIX), The Open Group Base Specifications Issue 6, Copyright (C) 2001-2004 by the Institute of Electrical and Electronics Engineers, Inc and The Open Group. In the event of any discrepancy between these versions and the original IEEE and The Open Group Standard, the original IEEE and The Open Group Standard is the referee document. The original Standard can be obtained online at http://www.opengroup.org/unix/online.html.  This notice shall appear on any product containing this material. 
.TH time 1 "1 Feb 1995" "SunOS 5.11" "User Commands"
.SH NAME
time \- time a simple command
.SH SYNOPSIS
.LP
.nf
\fBtime\fR [\fB-p\fR] \fIutility\fR [\fIargument\fR]...
.fi

.SH DESCRIPTION
.sp
.LP
The \fBtime\fR utility invokes \fIutility\fR operand with \fIargument\fR, and writes a message to standard error that lists timing statistics for \fIutility\fR. The message includes the following information:
.RS +4
.TP
.ie t \(bu
.el o
The elapsed (real) time between invocation of \fIutility\fR and its termination.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The User \fBCPU\fR time, equivalent to the sum of the \fItms_utime\fR and \fItms_cutime\fR fields returned by the \fBtimes\fR(2) function for the process in which \fIutility\fR is executed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The System \fBCPU\fR time, equivalent to the sum of the \fItms_stime\fR and \fItms_cstime\fR fields returned by the \fBtimes()\fR function for the process in which \fIutility\fR is executed.
.RE
.sp
.LP
When \fBtime\fR is used as part of a pipeline, the times reported are unspecified, except when it is the sole command within a grouping command in that pipeline. For example, the commands on the left are unspecified; those on the right report on utilities \fBa\fR and \fBc\fR, respectively:
.sp
.in +2
.nf
time a | b | c		{ time a } | b | c
a | b | time c		a | b | (time c)
.fi
.in -2

.SH OPTIONS
.sp
.LP
The following option is supported:
.sp
.ne 2
.mk
.na
\fB\fB-p\fR \fR
.ad
.RS 7n
.rt  
Writes the timing output to standard error in the following format:
.sp
.in +2
.nf
real %f\enuser %f\ensys %f\en < \fIreal seconds\fR>, <\fIuser seconds\fR>,
<\fIsystem seconds\fR>
.fi
.in -2
.sp

.RE

.SH OPERANDS
.sp
.LP
The following operands are supported:
.sp
.ne 2
.mk
.na
\fB\fIutility\fR \fR
.ad
.RS 13n
.rt  
The name of the utility that is to be invoked.
.RE

.sp
.ne 2
.mk
.na
\fB\fIargument\fR \fR
.ad
.RS 13n
.rt  
Any string to be supplied as an argument when invoking \fIutility\fR.
.RE

.SH USAGE
.sp
.LP
The \fBtime\fR utility returns exit status \fB127\fR if an error occurs so that applications can distinguish "failure to find a utility" from "invoked utility exited with an error indication." The value \fB127\fR was chosen because it is not commonly used for other meanings. Most utilities use small values for "normal error conditions" and the values above \fB128\fR can be confused with termination due to receipt of a signal. The value \fB126\fR was chosen in a similar manner to indicate that the utility could be found, but not invoked.
.SH EXAMPLES
.LP
\fBExample 1 \fRUsing the time command
.sp
.LP
It is frequently desirable to apply \fBtime\fR to pipelines or lists of commands. This can be done by placing pipelines and command lists in a single file. This single file can then be invoked as a utility, and the \fBtime\fR applies to everything in the file.

.sp
.LP
Alternatively, the following command can be used to apply \fBtime\fR to a complex command:

.sp
.in +2
.nf
example% \fBtime sh -c 'complex-command-line'\fR
.fi
.in -2
.sp

.LP
\fBExample 2 \fRUsing time in the csh shell
.sp
.LP
The following two examples show the differences between the \fBcsh\fR version of \fBtime\fR and the version in \fB/usr/bin/time\fR. These examples assume that \fBcsh\fR is the shell in use.

.sp
.in +2
.nf
example% \fBtime find / -name csh.1 -print
/usr/share/man/man1/csh.1\fR
95.0u 692.0s 1:17:52 16% 0+0k 0+0io 0pf+0w
.fi
.in -2
.sp

.sp
.LP
See \fBcsh\fR(1) for an explanation of the format of \fBtime\fR output.

.sp
.in +2
.nf
example% \fB/usr/bin/time find / -name csh.1 -print
/usr/share/man/man1/csh.1\fR
real  1:23:31.5
user     1:33.2
sys     11:28.2
.fi
.in -2
.sp

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(5) for descriptions of the following environment variables that affect the execution of \fBtime\fR: \fBLANG\fR, \fBLC_ALL\fR, \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, \fBLC_NUMERIC\fR,  \fBNLSPATH\fR, and \fBPATH\fR. 
.SH EXIT STATUS
.sp
.LP
If \fIutility\fR is invoked, the exit status of \fBtime\fR will be the exit status of \fIutility\fR. Otherwise, the \fBtime\fR utility will exit with one of the following values:
.sp
.ne 2
.mk
.na
\fB\fB1\(mi125\fR \fR
.ad
.RS 13n
.rt  
An error occurred in the \fBtime\fR utility.
.RE

.sp
.ne 2
.mk
.na
\fB\fB126\fR \fR
.ad
.RS 13n
.rt  
\fIutility\fR was found but could not be invoked.
.RE

.sp
.ne 2
.mk
.na
\fB\fB127\fR \fR
.ad
.RS 13n
.rt  
\fIutility\fR could not be found.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
_
Interface StabilityCommitted
_
StandardSee \fBstandards\fR(5).
.TE

.SH SEE ALSO
.sp
.LP
\fBcsh\fR(1), \fBshell_builtins\fR(1), \fBtimex\fR(1), \fBtimes\fR(2), \fBattributes\fR(5), \fBenviron\fR(5), \fBstandards\fR(5)
.SH NOTES
.sp
.LP
When the time command is run on a multiprocessor machine, the total of the values printed for \fBuser\fR and \fBsys\fR can exceed \fBreal\fR. This is  because on a multiprocessor machine it is possible to divide the task between the various processors.
.sp
.LP
When the command being timed is interrupted, the timing values displayed may not always be accurate.
.SH BUGS
.sp
.LP
Elapsed time is accurate to the second, while the \fBCPU\fR times are measured to the 100th second. Thus the sum of the \fBCPU\fR times can be up to a second larger than the elapsed time.


usr/src/cmd/time/time.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 1988, 2010, Oracle and/or its affiliates. All rights reserved.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*
 *	Time a command
 */

#include	<stdio.h>
#include	<signal.h>
#include	<errno.h>
#include	<stdlib.h>
#include	<unistd.h>
#include	<libintl.h>
#include	<locale.h>
#include	<limits.h>
#include	<sys/types.h>
#include	<sys/times.h>
#include	<sys/wait.h>

/*
 * The following use of HZ/10 will work correctly only if HZ is a multiple
 * of 10.  However the only values for HZ now in use are 100 for the 3B
 * and 60 for other machines.
 *
 * The first value was HZ/10. Since HZ should be gotten from sysconf()
 * it is dynamically initialized at entry to the main program.
 */
static clock_t quant[] = { 10, 10, 10, 6, 10, 6, 10, 10, 10 };
static char *pad  = "000      ";
static char *sep  = "\0\0.\0:\0:\0\0";
static char *nsep = "\0\0.\0 \0 \0\0";

static void usage(void);
static void printt(char *, clock_t);

int
main(int argc, char **argv)
{
	struct tms	buffer;
	pid_t		p;
	int		status;
	int		pflag		= 0;
	int		c;
	int		clock_tick	= CLK_TCK;
	clock_t		before, after;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN "SYS_TEST"
#endif
	(void) textdomain(TEXT_DOMAIN);

	while ((c = getopt(argc, argv, "p")) != EOF)
		switch (c) {
		case 'p':
			pflag++;
			break;
		case '?':
			usage();
		}

	argc -= optind;
	argv += optind;

	/*
	 * time(1) is only accurate to a tenth of a second.  We need to
	 * determine the number of clock ticks in a tenth of a second in
	 * order to later divide away what we don't care about.
	 */
	quant[0] = clock_tick/10;

	before = times(&buffer);
	if (argc < 1)
		usage();
	p = fork();
	if (p == (pid_t)-1) {
		perror("time");
		exit(2);
	}
	if (p == (pid_t)0) {
		(void) execvp(argv[0], &argv[0]);
		perror(argv[0]);
		if (errno == ENOENT)
			exit(127);
		else
			exit(126);
	}
	(void) signal(SIGINT, SIG_IGN);
	(void) signal(SIGQUIT, SIG_IGN);
	while (waitpid(p, &status, 0) == -1 && errno == EINTR)
		;
	if ((status & 0377) != '\0')
		(void) fprintf(stderr, "time: %s\n",
		    gettext("command terminated abnormally."));
	after = times(&buffer);
	(void) fprintf(stderr, "\n");
	if (pflag)
		(void) fprintf(stderr, "real %.2f\nuser %.2f\nsys %.2f\n",
		    (double)(after-before)/clock_tick,
		    (double)buffer.tms_cutime/clock_tick,
		    (double)buffer.tms_cstime/clock_tick);
	else {
		printt("real", (after-before));
		printt("user", buffer.tms_cutime);
		printt("sys ", buffer.tms_cstime);
	}

	return ((status & 0xff00)
	    ? (status >> 8)
	    : ((status & 0x00ff) ? ((status & ~WCOREFLG) | 0200) : 0));
}


static void
printt(char *s, clock_t a)
{
	int i;
	char digit[9];
	char c;
	int nonzero;

	a /= quant[0];	/* Divide away the accuracy we don't care about */

	/*
	 * We now have the number of tenths of seconds elapsed in terms of
	 * ticks. Loop through to determine the actual digits.
	 */
	for (i = 1; i < 9; i++) {
		digit[i] = a % quant[i];
		a /= quant[i];
	}
	(void) fprintf(stderr, "%s", s);
	nonzero = 0;
	while (--i > 0) {
		c = (digit[i] != 0) ? digit[i]+'0' : (nonzero ? '0': pad[i]);
		if (c != '\0')
			(void) putc(c, stderr);
		nonzero |= digit[i];
		c = nonzero?sep[i]:nsep[i];
		if (c != '\0')
			(void) putc(c, stderr);
	}
	(void) fprintf(stderr, "\n");
}

static void
usage(void)
{
	(void) fprintf(stderr,
	    gettext("usage: time [-p] utility [argument...]\n"));
	exit(1);
}
